### PR TITLE
WIP: Provide method addition

### LINKED ISSUE #477: Add an implementation for provide in HttpApp and downstream types
* Add `provide` method in HttpApp and downstream types
* Add unit test cases for `provide` in HttpApp

### previously titled: Fix provide Method and Add Unit tests for `provide` in `HttpApp`

### PR DESCRIPTION
**Pull request by girdharshubham**
# Changelog
* provide method added for
  * Socket
  * SocketApp
  * HttpAttribute
  * Response
  * HttpApp
* Unit test case added for Socket's provide
* Unit test case added for HttpApp's provide
### Closes #477

### CODE REVIEW

File: zio-http/src/main/scala/zhttp/http/HttpAttribute.scala [hunk 1]
```diff
@@ -1,13 +1,20 @@
...
 import zhttp.socket.SocketApp
+import zio.NeedsEnv
 
 /**
  * Attribute holder for and Responses
  */
 private[zhttp] sealed trait HttpAttribute[-R, +E] extends Product with Serializable { self =>
   def exit: HttpAttribute.AttributeExit[R, E]                                   = HttpAttribute.exit(self)
   def ++[R1 <: R, E1 >: E](other: HttpAttribute[R1, E1]): HttpAttribute[R1, E1] = HttpAttribute.Combine(self, other)
+  def provide(r: R)(implicit env: NeedsEnv[R]): HttpAttribute[Any, E]           = self match {
```
Comment (tusharmath) [inline zio-http/src/main/scala/zhttp/http/HttpAttribute.scala:13]: Create a new type for `HttpAttribute.Provide`

File: zio-http/src/main/scala/zhttp/socket/Socket.scala [hunk 2]
```diff
@@ -20,6 +20,9 @@ sealed trait Socket[-R, +E, -A, +B] { self =>
 
   private[zhttp] def execute(a: A): ZStream[R, E, B] = self(a)
 
+  def provide(r: R)(implicit needsEnv: NeedsEnv[R]): Socket[Any, E, A, B] =
+    Socket.FromStreamingFunction(a => self(a).provide(r))
```
Comment (tusharmath) [inline zio-http/src/main/scala/zhttp/socket/Socket.scala]:
```suggestion
    Socket.Provide(r, self)
```

File: zio-http/src/main/scala/zhttp/socket/SocketApp.scala [hunk 3]
```diff
@@ -49,6 +50,23 @@ sealed trait SocketApp[-R, +E] { self =>
     loop(self, SocketConfig[R, E]())
   }
 
+  def provide(r: R)(implicit env: NeedsEnv[R]): SocketApp[Any, E] =
```
Comment (tusharmath) [inline zio-http/src/main/scala/zhttp/socket/SocketApp.scala:53]:
```suggestion
  def provide(r: R)(implicit env: NeedsEnv[R]): SocketApp[Any, E] = Provide(r, self)
```

File: zio-http/src/main/scala/zhttp/socket/Socket.scala (hunk 2 above)
Comment (girdharshubham) [reply zio-http/src/main/scala/zhttp/socket/Socket.scala]: While providing an implementation for the Provide type and the method, the compiler keeps on warning that the effect has **Any** as its requirement
```scala
sealed trait Socket[-R, +E, -A, +B] { self =>
  def apply(a: A): ZStream[R, E, B] = self match {
    ...
    case Provide(m, r)               => m(a).provide(r)
  }
  def provide(r: R)(implicit needsEnv: NeedsEnv[R]): Socket[Any, E, A, B] = Provide(self, r)
}
object Socket {
  private final case class Provide[R, E, A, B](m: Socket[R, E, A, B], r: R)                 extends Socket[Any, E, A, B]
}
```

I believe the intention behind your comment is that with provide we should return a declaration of what we'd end up with and then the apply method is where the actual environment elimination should be.

Comment (girdharshubham) [reply zio-http/src/main/scala/zhttp/socket/Socket.scala]: This has been resolved in newer commit(s)

**Comment by girdharshubham** [comment]
Closing this PR in favour of smaller scoped PRs.
Comments will be resolved in them.

**Pull request closed without merging**